src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+        for (String microarrayType : microarrayTypes) {
+            try {
+                logger.info("Retrieving data associated with microarray type " + microarrayType);
+                Set<String> microarrayResponseLines = queryBiomart(getBiomartIdentifierQuery(biomartXMLPath, biomartSpeciesName, microarrayType), 0);
+                Set<String> microarrayIdentifierLines = new HashSet<>();
+                for (String microarrayResponseLine : microarrayResponseLines) {
+                    List<String> tabSplit = Arrays.asList(microarrayResponseLine.split("\t"));
+                    // Microarray ID is 4th column, and sometimes it isn't in a line.
+                    // Since we need the ID, we require the List to have 4 elements.
+                    if (tabSplit.size() == 4) {
+                        microarrayIdentifierLines.add(microarrayResponseLine + "\n");
+                    }
+                }
+
+                // Write data to file.
+                if (!microarrayIdentifierLines.isEmpty()) {
+                    File speciesMicroarrayFile = createNewFile(this.destination + biomartSpeciesName + "_microarray");
+                    for (String probeIdLine : microarrayIdentifierLines) {
+                        Files.write(Paths.get(speciesMicroarrayFile.toString()), probeIdLine.getBytes(), StandardOpenOption.APPEND);
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Attempts to create a connection to Biomart with the supplied query
+     * @param queryString String of Biomart query
+     * @return Set<String> Biomart response data.
+     * @throws Exception Can be caused by the query URL not being valid, or the data not existing for the species.
+     */
+    private Set<String> queryBiomart(String queryString, int retryCount) throws Exception {
+        // Create connection to Biomart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == 200) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                // Biomart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    retryCount++;
+                    if (retryCount < 5) {
+                        Thread.sleep(10000);
```

Comment:
10000 can be extracted to a constant. OR you could use the `Duration` class if you want something that's easier to read. You could replace 10000 with `Duration.of(10, SECONDS)` (and I think you'd need to append that with `.toMillis()`.

Reply:
5 can also be extracted to a named constant such as `MAX_QUERY_RETRIES`